### DISCUSSION TITLE
(at least partial) Firmware, Power Management and Other GPU Documentation

### DISCUSSION THREAD

**Discussion by ghost** [NVIDIA/open-gpu-kernel-modules#210] [General]
For extra enhancements and bug fixes regarding specific gpu functions such as Power Management, Firmware ABI, etc. we will need actual documentation of these features. Please provide as much documentation as possible.

**Comment by PAR2020** [comment]
Thanks for the request, and vote up. I'm going to move this issue to Discussions to get some more visibility since it is not a bug or an enhancement.

**Comment by YusufKhan-gamedev** [comment]
Add memory management, performance counter, mmio, reclocking, and PDAEMON documentation to the list

**Comment by YusufKhan-gamedev** [comment]
https://nvidia.github.io/open-gpu-doc/ Already has some documentation, extend that.

**Reply by mtijanic** [reply]
For what it's worth, there are plenty more [reference manuals](https://github.com/NVIDIA/open-gpu-kernel-modules/tree/main/src/common/inc/swref/published) and [class headers](https://github.com/NVIDIA/open-gpu-kernel-modules/tree/main/src/common/sdk/nvidia/inc/class) in this repository that still haven't made their way into `open-gpu-doc`.

**Comment by jeremyyx** [comment]
vote up!

**Comment by Ved-un** [comment]
vote up!

**Comment by princesstwi** [comment]
Vote up!

**Reply by aaronp24** [reply]
Please use the upvote button at the top rather than posting "vote up" replies.

**Reply by princesstwi** [reply]
I did both

**Reply by YusufKhan-gamedev** [reply]
lol

> I did both

**Reply by mobin-2008** [reply]
> I did both

bruh

**Comment by LupertEverett** [comment]
Seconding this for especially older cards. Even though Nvidia wouldn't support these cards with their own driver, doing this and letting the community handle the rest could be of an immense help.

**Reply by YusufKhan-gamedev** [reply]
Isnt nouveau competent for older cards already? Just not maxwell/pascal as they dont have reclocking firmware

**Reply by LupertEverett** [reply]
To my knowledge (aka whatever I was able to dig up from nouveau's site), dynamic reclocking and power management is still a wip even with older cards. That's one of the reasons I am not using nouveau even though I have a Kepler card, aside from the lack of Vulkan support.

**Reply by YusufKhan-gamedev** [reply]
Even though its reclocking is marked as WIP its mature enough that you could go along with your life wih it now...Also vulkan stubs are public now! It doesn't even fully support Vulkan 1.0 but its getting there...I have [a fork](https://gitlab.freedesktop.org/YusufKhan-gamedev/mesa/-/commits/vulkan/) that can be used to compile it and a local fork that adds support for ogk.

https://gitlab.freedesktop.org/karolherbst/mesa/-/tree/nouveau/push/src/nouveau

**Reply by LupertEverett** [reply]
Is that so? That's good news then! I think I should give nouveau another try soon! Thanks for the heads up!

**Reply by YusufKhan-gamedev** [reply]
Please note that the vulkan driver is *just* getting started and is no where near developed as intel, amd or even the arm mali phones vulkan implementation hence why it isnt in Mesa yet. 

For extra performance you might want to see a [merge request](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10752) by Karol Herbst which enables multi-threading, the MR direly needs testing on more hardware and distros.

**Reply by edisionnano** [reply]
Vulkan Driver? What Vulkan Driver?

**Reply by YusufKhan-gamedev** [reply]
nvk, its public now but not upstream, see https://gitlab.freedesktop.org/karolherbst/mesa/-/tree/nouveau/push/src/nouveau

**Reply by edisionnano** [reply]
thanks

**Reply by YusufKhan-gamedev** [reply]
You need to patch it if you want to use it, apply the patch from my fork to it and you'll be fine

**Reply by LupertEverett** [reply]
> Please note that the vulkan driver is _just_ getting started and is no where near developed as intel, amd or even the arm mali phones vulkan implementation hence why it isnt in Mesa yet.
> 
> For extra performance you might want to see a [merge request](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10752) by Karol Herbst which enables multi-threading, the MR direly needs testing on more hardware and distros.

So I gave a shot at nouveau today. After some efforts, I was able to uninstall the nvidia driver and remove the blacklisting on nouveau. Did a quick test with some old games and while they worked alright-ish, there were some slowdowns that I didn't experience with the 470.xx drivers

The bottom line is, I am willing to test those performance improvements out, but I need instructions on how to install it on my system, as I am not really experienced with these.

**Reply by YusufKhan-gamedev** [reply]
If you want to download the source code for the mesa multithreading patches run: 

```
git clone https://gitlab.freedesktop.org/karolherbst/mesa.git
cd mesa
git checkout nvc0_threading_upstream
```
Then follow https://docs.mesa3d.org/install.html for compiling and installing the patches, the 3rd step and beyond is NOT needed.

With the patches I saw a 20% performance boost.

**Reply by edisionnano** [reply]
The NVK code moved to https://gitlab.freedesktop.org/nouveau/mesa/-/commits/nouveau/vk/

**Reply by YusufKhan-gamedev** [reply]
To enable that patch instead of the multithreading support(please note your sacrificing performance for features when you do this) you run git checkout `nouveau/vk` on the mesa directory and then follow the mesa build instructions as provided.

**Reply by YusufKhan-gamedev** [reply]
Also if you notice any bugs with the multi-threading patches be sure to report them! See the merge request for more info.

**Reply by LupertEverett** [reply]
Sadly it didn't have any performance increases on my end, at least with the stuff I was able to test it so far. Provided my feedback there.

Hope I did it correctly.

**Reply by YusufKhan-gamedev** [reply]
`glxinfo | grep OpenGL` ? It could be possible(likely) ninja got confused somewhere and installed things in the wrong place.

**Reply by LupertEverett** [reply]
> `glxinfo | grep OpenGL` ? It could be possible(likely) ninja got confused somewhere and installed things in the wrong place.

Tbh I tested it with `DRI_PRIME=1 glxgears -info` (due to optimus), and every time I did that, it reported the mesa version as the stable version from the repos, even after following the installation instructions carefully. I had to do a "local installation", then manually point to that installation dir to ensure that I'm indeed running the newer code. (And that's how I did the "before/after" tests.) Before that, I was wondering why I wasn't noticing any differences, and it took me pretty long to figure this all out lol.

(Also ngl, I think this convo has gone quite off-topic. Should we move it to somewhere else, if possible?)

**Reply by YusufKhan-gamedev** [reply]
*sigh* ninja bug...I dont know much you can do with it except file a issue on the ninja github page and see what they say

or you can just forget it and go back to the proprietary driver and wait for these changes to be upstream in your distro then switch to nouveau ;)

**Reply by LupertEverett** [reply]
> _sigh_ ninja bug...I dont know much you can do with it except file a issue on the ninja github page and see what they say

Yeah that makes the most sense, but I am kinda fine with this setup lol, as it allows me to test out the new stuff without messing up with my system.

> or you can just forget it and go back to the proprietary driver and wait for these changes to be upstream in your distro then switch to nouveau ;)

Ngl this is also a tempting option :V

I might stay around a lil bit first though :)

At least I am glad that things seem to be better than I thought them to be.

**Reply by YusufKhan-gamedev** [reply]
A big problem is that a lot of stuff is downstream...I have a draft MR to *try* to get stuff upstream but...

**Reply by LupertEverett** [reply]
I wish you success with that, honestly.
